README
NAME
    Scope::Upper - Act on upper scopes.

VERSION
    Version 0.28

SYNOPSIS
    "reap", "localize", "localize_elem", "localize_delete" and "WORDS" :

        package Scope;

        use Scope::Upper qw<
         reap localize localize_elem localize_delete
         :words
        >;

        sub new {
         my ($class, $name) = @_;

         localize '$tag' => bless({ name => $name }, $class) => UP;

         reap { print Scope->tag->name, ": end\n" } UP;
        }

        # Get the tag stored in the caller namespace
        sub tag {
         my $l   = 0;
         my $pkg = __PACKAGE__;
         $pkg    = caller $l++ while $pkg eq __PACKAGE__;

         no strict 'refs';
         ${$pkg . '::tag'};
        }

        sub name { shift->{name} }

        # Locally capture warnings and reprint them with the name prefixed
        sub catch {
         localize_elem '%SIG', '__WARN__' => sub {
          print Scope->tag->name, ': ', @_;
         } => UP;
        }

        # Locally clear @INC
        sub private {
         for (reverse 0 .. $#INC) {
          # First UP is the for loop, second is the sub boundary
          localize_delete '@INC', $_ => UP UP;
         }
        }

        ...

        package UserLand;

        {
         Scope->new("top");    # initializes $UserLand::tag

         {
          Scope->catch;
          my $one = 1 + undef; # prints "top: Use of uninitialized value..."

          {
           Scope->private;
           eval { require Cwd };
           print $@;           # prints "Can't locate Cwd.pm in @INC
          }                    #         (@INC contains:) at..."

          require Cwd;         # loads Cwd.pm
         }

        }                      # prints "top: done"

    "unwind" and "want_at" :

        package Try;

        use Scope::Upper qw<unwind want_at :words>;

        sub try (&) {
         my @result = shift->();
         my $cx = SUB UP; # Point to the sub above this one
         unwind +(want_at($cx) ? @result : scalar @result) => $cx;
        }

        ...

        sub zap {
         try {
          my @things = qw<a b c>;
          return @things; # returns to try() and then outside zap()
          # not reached
         };
         # not reached
        }

        my @stuff = zap(); # @stuff contains qw<a b c>
        my $stuff = zap(); # $stuff contains 3

    "uplevel" :

        package Uplevel;

        use Scope::Upper qw<uplevel CALLER>;

        sub target {
         faker(@_);
        }

        sub faker {
         uplevel {
          my $sub = (caller 0)[3];
          print "$_[0] from $sub()";
         } @_ => CALLER(1);
        }

        target('hello'); # "hello from Uplevel::target()"

    "uid" and "validate_uid" :

        use Scope::Upper qw<uid validate_uid>;

        my $uid;

        {
         $uid = uid();
         {
          if ($uid eq uid(UP)) { # yes
           ...
          }
          if (validate_uid($uid)) { # yes
           ...
          }
         }
        }

        if (validate_uid($uid)) { # no
         ...
        }

DESCRIPTION
    This module lets you defer actions *at run-time* that will take place
    when the control flow returns into an upper scope. Currently, you can:

    *   hook an upper scope end with "reap" ;

    *   localize variables, array/hash values or deletions of elements in
        higher contexts with respectively "localize", "localize_elem" and
        "localize_delete" ;

    *   return values immediately to an upper level with "unwind", "yield"
        and "leave" ;

    *   gather information about an upper context with "want_at" and
        "context_info" ;

    *   execute a subroutine in the setting of an upper subroutine stack
        frame with "uplevel" ;

    *   uniquely identify contexts with "uid" and "validate_uid".

FUNCTIONS
    In all those functions, $context refers to the target scope.

    You have to use one or a combination of "WORDS" to build the $context
    passed to these functions. This is needed in order to ensure that the
    module still works when your program is ran in the debugger. The only
    thing you can assume is that it is an *absolute* indicator of the frame,
    which means that you can safely store it at some point and use it when
    needed, and it will still denote the original scope.

  "reap"
        reap { ... };
        reap { ... } $context;
        &reap($callback, $context);

    Adds a destructor that calls $callback (in void context) when the upper
    scope represented by $context ends.

  "localize"
        localize $what, $value;
        localize $what, $value, $context;

    Introduces a "local" delayed to the time of first return into the upper
    scope denoted by $context. $what can be :

    *   A glob, in which case $value can either be a glob or a reference.
        "localize" follows then the same syntax as "local *x = $value". For
        example, if $value is a scalar reference, then the "SCALAR" slot of
        the glob will be set to $$value - just like "local *x = \1" sets $x
        to 1.

    *   A string beginning with a sigil, representing the symbol to localize
        and to assign to. If the sigil is '$', "localize" follows the same
        syntax as "local $x = $value", i.e. $value isn't dereferenced. For
        example,

            localize '$x', \'foo' => HERE;

        will set $x to a reference to the string 'foo'. Other sigils ('@',
        '%', '&' and '*') require $value to be a reference of the
        corresponding type.

        When the symbol is given by a string, it is resolved when the actual
        localization takes place and not when "localize" is called. Thus, if
        the symbol name is not qualified, it will refer to the variable in
        the package where the localization actually takes place and not in
        the one where the "localize" call was compiled. For example,

            {
             package Scope;
             sub new { localize '$tag', $_[0] => UP }
            }

            {
             package Tool;
             {
              Scope->new;
              ...
             }
            }

        will localize $Tool::tag and not $Scope::tag. If you want the other
        behaviour, you just have to specify $what as a glob or a qualified
        name.

        Note that if $what is a string denoting a variable that wasn't
        declared beforehand, the relevant slot will be vivified as needed
        and won't be deleted from the glob when the localization ends. This
        situation never arises with "local" because it only compiles when
        the localized variable is already declared. Although I believe it
        shouldn't be a problem as glob slots definedness is pretty much an
        implementation detail, this behaviour may change in the future if
        proved harmful.

  "localize_elem"
        localize_elem $what, $key, $value;
        localize_elem $what, $key, $value, $context;

    Introduces a "local $what[$key] = $value" or "local $what{$key} =
    $value" delayed to the time of first return into the upper scope denoted
    by $context. Unlike "localize", $what must be a string and the type of
    localization is inferred from its sigil. The two only valid types are
    array and hash ; for anything besides those, "localize_elem" will throw
    an exception. $key is either an array index or a hash key, depending of
    which kind of variable you localize.

    If $what is a string pointing to an undeclared variable, the variable
    will be vivified as soon as the localization occurs and emptied when it
    ends, although it will still exist in its glob.

  "localize_delete"
        localize_delete $what, $key;
        localize_delete $what, $key, $context;

    Introduces the deletion of a variable or an array/hash element delayed
    to the time of first return into the upper scope denoted by $context.
    $what can be:

    *   A glob, in which case $key is ignored and the call is equivalent to
        "local *x".

    *   A string beginning with '@' or '%', for which the call is equivalent
        to respectively "local $a[$key]; delete $a[$key]" and "local
        $h{$key}; delete $h{$key}".

    *   A string beginning with '&', which more or less does "undef &func"
        in the upper scope. It's actually more powerful, as &func won't even
        "exists" anymore. $key is ignored.

  "unwind"
        unwind;
        unwind @values, $context;

    Returns @values *from* the subroutine, eval or format context pointed by
    or just above $context, and immediately restarts the program flow at
    this point - thus effectively returning @values to an upper scope. If
    @values is empty, then the $context parameter is optional and defaults
    to the current context (making the call equivalent to a bare "return;")
    ; otherwise it is mandatory.

    The upper context isn't coerced onto @values, which is hence always
    evaluated in list context. This means that

        my $num = sub {
         my @a = ('a' .. 'z');
         unwind @a => HERE;
         # not reached
        }->();

    will set $num to 'z'. You can use "want_at" to handle these cases.

  "yield"
        yield;
        yield @values, $context;

    Returns @values *from* the context pointed by or just above $context,
    and immediately restarts the program flow at this point. If @values is
    empty, then the $context parameter is optional and defaults to the
    current context ; otherwise it is mandatory.

    "yield" differs from "unwind" in that it can target *any* upper scope
    (besides a "s///e" substitution context) and not necessarily a sub, an
    eval or a format. Hence you can use it to return values from a "do" or a
    "map" block :

        my $now = do {
         local $@;
         eval { require Time::HiRes } or yield time() => HERE;
         Time::HiRes::time();
        };

        my @uniq = map {
         yield if $seen{$_}++; # returns the empty list from the block
         ...
        } @things;

    Like for "unwind", the upper context isn't coerced onto @values. You can
    use the fifth value returned by "context_info" to handle context
    coercion.

  "leave"
        leave;
        leave @values;

    Immediately returns @values from the current block, whatever it may be
    (besides a "s///e" substitution context). "leave" is actually a synonym
    for "yield HERE", while "leave @values" is a synonym for "yield @values,
    HERE".

    Like for "yield", you can use the fifth value returned by "context_info"
    to handle context coercion.

  "want_at"
        my $want = want_at;
        my $want = want_at $context;

    Like "wantarray" in perlfunc, but for the subroutine, eval or format
    context located at or just above $context.

    It can be used to revise the example showed in "unwind" :

        my $num = sub {
         my @a = ('a' .. 'z');
         unwind +(want_at(HERE) ? @a : scalar @a) => HERE;
         # not reached
        }->();

    will rightfully set $num to 26.

  "context_info"
        my ($package, $filename, $line, $subroutine, $hasargs,
            $wantarray, $evaltext, $is_require, $hints, $bitmask,
            $hinthash) = context_info $context;

    Gives information about the context denoted by $context, akin to what
    "caller" in perlfunc provides but not limited only to subroutine, eval
    and format contexts. When $context is omitted, it defaults to the
    current context.

    The returned values are, in order :

    *   *(index 0)* : the namespace in use when the context was created ;

    *   *(index 1)* : the name of the file at the point where the context
        was created ;

    *   *(index 2)* : the line number at the point where the context was
        created ;

    *   *(index 3)* : the name of the subroutine called for this context, or
        "undef" if this is not a subroutine context ;

    *   *(index 4)* : a boolean indicating whether a new instance of @_ was
        set up for this context, or "undef" if this is not a subroutine
        context ;

    *   *(index 5)* : the context (in the sense of "wantarray" in perlfunc)
        in which the context (in our sense) is executed ;

    *   *(index 6)* : the contents of the string being compiled for this
        context, or "undef" if this is not an eval context ;

    *   *(index 7)* : a boolean indicating whether this eval context was
        created by "require", or "undef" if this is not an eval context ;

    *   *(index 8)* : the value of the lexical hints in use when the context
        was created ;

    *   *(index 9)* : a bit string representing the warnings in use when the
        context was created ;

    *   *(index 10)* : a reference to the lexical hints hash in use when the
        context was created (only on perl 5.10 or greater).

  "uplevel"
        my @ret = uplevel { ...; return @ret };
        my @ret = uplevel { my @args = @_; ...; return @ret } @args, $context;
        my @ret = &uplevel($callback, @args, $context);

    Executes the code reference $callback with arguments @args as if it were
    located at the subroutine stack frame pointed by $context, effectively
    fooling "caller" and "die" into believing that the call actually
    happened higher in the stack. The code is executed in the context of the
    "uplevel" call, and what it returns is returned as-is by "uplevel".

        sub target {
         faker(@_);
        }

        sub faker {
         uplevel {
          map { 1 / $_ } @_;
         } @_ => CALLER(1);
        }

        my @inverses = target(1, 2, 4); # @inverses contains (0, 0.5, 0.25)
        my $count    = target(1, 2, 4); # $count is 3

    Note that if @args is empty, then the $context parameter is optional and
    defaults to the current context ; otherwise it is mandatory.

    Sub::Uplevel also implements a pure-Perl version of "uplevel". Both are
    identical, with the following caveats :

    *   The Sub::Uplevel implementation of "uplevel" may execute a code
        reference in the context of any upper stack frame. The Scope::Upper
        version can only uplevel to a subroutine stack frame, and will croak
        if you try to target an "eval" or a format.

    *   Exceptions thrown from the code called by this version of "uplevel"
        will not be caught by "eval" blocks between the target frame and the
        uplevel call, while they will for Sub::Uplevel's version. This means
        that :

            eval {
             sub {
              local $@;
              eval {
               sub {
                uplevel { die 'wut' } CALLER(2); # for Scope::Upper
                # uplevel(3, sub { die 'wut' })  # for Sub::Uplevel
               }->();
              };
              print "inner block: $@";
              $@ and exit;
             }->();
            };
            print "outer block: $@";

        will print "inner block: wut..." with Sub::Uplevel and "outer block:
        wut..." with Scope::Upper.

    *   Sub::Uplevel globally overrides the Perl keyword "caller", while
        Scope::Upper does not.

    A simple wrapper lets you mimic the interface of "uplevel" in
    Sub::Uplevel :

        use Scope::Upper;

        sub uplevel {
         my $frame = shift;
         my $code  = shift;
         my $cxt   = Scope::Upper::CALLER($frame);
         &Scope::Upper::uplevel($code => @_ => $cxt);
        }

    Albeit the three exceptions listed above, it passes all the tests of
    Sub::Uplevel.

  "uid"
        my $uid = uid;
        my $uid = uid $context;

    Returns an unique identifier (UID) for the context (or dynamic scope)
    pointed by $context, or for the current context if $context is omitted.
    This UID will only be valid for the life time of the context it
    represents, and another UID will be generated next time the same scope
    is executed.

        my $uid;

        {
         $uid = uid;
         if ($uid eq uid()) { # yes, this is the same context
          ...
         }
         {
          if ($uid eq uid()) { # no, we are one scope below
           ...
          }
          if ($uid eq uid(UP)) { # yes, UP points to the same scope as $uid
           ...
          }
         }
        }

        # $uid is now invalid

        {
         if ($uid eq uid()) { # no, this is another block
          ...
         }
        }

    For example, each loop iteration gets its own UID :

        my %uids;

        for (1 .. 5) {
         my $uid = uid;
         $uids{$uid} = $_;
        }

        # %uids has 5 entries

    The UIDs are not guaranteed to be numbers, so you must use the "eq"
    operator to compare them.

    To check whether a given UID is valid, you can use the "validate_uid"
    function.

  "validate_uid"
        my $is_valid = validate_uid $uid;

    Returns true if and only if $uid is the UID of a currently valid context
    (that is, it designates a scope that is higher than the current one in
    the call stack).

        my $uid;

        {
         $uid = uid();
         if (validate_uid($uid)) { # yes
          ...
         }
         {
          if (validate_uid($uid)) { # yes
           ...
          }
         }
        }

        if (validate_uid($uid)) { # no
         ...
        }

CONSTANTS
  "SU_THREADSAFE"
    True iff the module could have been built when thread-safety features.

WORDS
  Constants
   "TOP"
        my $top_context = TOP;

    Returns the context that currently represents the highest scope.

   "HERE"
        my $current_context = HERE;

    The context of the current scope.

  Getting a context from a context
    For any of those functions, $from is expected to be a context. When
    omitted, it defaults to the current context.

   "UP"
        my $upper_context = UP;
        my $upper_context = UP $from;

    The context of the scope just above $from. If $from points to the
    top-level scope in the current stack, then a warning is emitted and
    $from is returned (see "DIAGNOSTICS" for details).

   "SUB"
        my $sub_context = SUB;
        my $sub_context = SUB $from;

    The context of the closest subroutine above $from. If $from already
    designates a subroutine context, then it is returned as-is ; hence "SUB
    SUB == SUB". If no subroutine context is present in the call stack, then
    a warning is emitted and the current context is returned (see
    "DIAGNOSTICS" for details).

   "EVAL"
        my $eval_context = EVAL;
        my $eval_context = EVAL $from;

    The context of the closest eval above $from. If $from already designates
    an eval context, then it is returned as-is ; hence "EVAL EVAL == EVAL".
    If no eval context is present in the call stack, then a warning is
    emitted and the current context is returned (see "DIAGNOSTICS" for
    details).

  Getting a context from a level
    Here, $level should denote a number of scopes above the current one.
    When omitted, it defaults to 0 and those functions return the same
    context as "HERE".

   "SCOPE"
        my $context = SCOPE;
        my $context = SCOPE $level;

    The $level-th upper context, regardless of its type. If $level points
    above the top-level scope in the current stack, then a warning is
    emitted and the top-level context is returned (see "DIAGNOSTICS" for
    details).

   "CALLER"
        my $context = CALLER;
        my $context = CALLER $level;

    The context of the $level-th upper subroutine/eval/format. It kind of
    corresponds to the context represented by "caller $level", but while
    e.g. "caller 0" refers to the caller context, "CALLER 0" will refer to
    the top scope in the current context. If $level points above the
    top-level scope in the current stack, then a warning is emitted and the
    top-level context is returned (see "DIAGNOSTICS" for details).

  Examples
    Where "reap" fires depending on the $cxt :

        sub {
         eval {
          sub {
           {
            reap \&cleanup => $cxt;
            ...
           }     # $cxt = SCOPE(0) = HERE
           ...
          }->(); # $cxt = SCOPE(1) = UP = SUB = CALLER(0)
          ...
         };      # $cxt = SCOPE(2) = UP UP =  UP SUB = EVAL = CALLER(1)
         ...
        }->();   # $cxt = SCOPE(3) = SUB UP SUB = SUB EVAL = CALLER(2)
        ...

    Where "localize", "localize_elem" and "localize_delete" act depending on
    the $cxt :

        sub {
         eval {
          sub {
           {
            localize '$x' => 1 => $cxt;
            # $cxt = SCOPE(0) = HERE
            ...
           }
           # $cxt = SCOPE(1) = UP = SUB = CALLER(0)
           ...
          }->();
          # $cxt = SCOPE(2) = UP UP = UP SUB = EVAL = CALLER(1)
          ...
         };
         # $cxt = SCOPE(3) = SUB UP SUB = SUB EVAL = CALLER(2)
         ...
        }->();
        # $cxt = SCOPE(4), UP SUB UP SUB = UP SUB EVAL = UP CALLER(2) = TOP
        ...

    Where "unwind", "yield", "want_at", "context_info" and "uplevel" point
    to depending on the $cxt:

        sub {
         eval {
          sub {
           {
            unwind @things => $cxt;   # or yield @things => $cxt
                                      # or uplevel { ... } $cxt
            ...
           }
           ...
          }->(); # $cxt = SCOPE(0) = SCOPE(1) = HERE = UP = SUB = CALLER(0)
          ...
         };      # $cxt = SCOPE(2) = UP UP = UP SUB = EVAL = CALLER(1) (*)
         ...
        }->();   # $cxt = SCOPE(3) = SUB UP SUB = SUB EVAL = CALLER(2)
        ...

        # (*) Note that uplevel() will croak if you pass that scope frame,
        #     because it cannot target eval scopes.

DIAGNOSTICS
  "Cannot target a scope outside of the current stack"
    This warning is emitted when "UP", "SCOPE" or "CALLER" end up pointing
    to a context that is above the top-level context of the current stack.
    It indicates that you tried to go higher than the main scope, or to
    point across a "DESTROY" method, a signal handler, an overloaded or tied
    method call, a "require" statement or a "sort" callback. In this case,
    the resulting context is the highest reachable one.

  "No targetable %s scope in the current stack"
    This warning is emitted when you ask for an "EVAL" or "SUB" context and
    no such scope can be found in the call stack. The resulting context is
    the current one.

EXPORT
    The functions "reap", "localize", "localize_elem", "localize_delete",
    "unwind", "yield", "leave", "want_at", "context_info" and "uplevel" are
    only exported on request, either individually or by the tags ':funcs'
    and ':all'.

    The constant "SU_THREADSAFE" is also only exported on request,
    individually or by the tags ':consts' and ':all'.

    Same goes for the words "TOP", "HERE", "UP", "SUB", "EVAL", "SCOPE" and
    "CALLER" that are only exported on request, individually or by the tags
    ':words' and ':all'.

CAVEATS
    It is not possible to act upon a scope that belongs to another perl
    'stack', i.e. to target a scope across a "DESTROY" method, a signal
    handler, an overloaded or tied method call, a "require" statement or a
    "sort" callback.

    Be careful that local variables are restored in the reverse order in
    which they were localized. Consider those examples:

        local $x = 0;
        {
         reap sub { print $x } => HERE;
         local $x = 1;
         ...
        }
        # prints '0'
        ...
        {
         local $x = 1;
         reap sub { $x = 2 } => HERE;
         ...
        }
        # $x is 0

    The first case is "solved" by moving the "local" before the "reap", and
    the second by using "localize" instead of "reap".

    The effects of "reap", "localize" and "localize_elem" can't cross
    "BEGIN" blocks, hence calling those functions in "import" is deemed to
    be useless. This is an hopeless case because "BEGIN" blocks are executed
    once while localizing constructs should do their job at each run.
    However, it's possible to hook the end of the current scope compilation
    with B::Hooks::EndOfScope.

    Some rare oddities may still happen when running inside the debugger. It
    may help to use a perl higher than 5.8.9 or 5.10.0, as they contain some
    context-related fixes.

    Calling "goto" to replace an "uplevel"'d code frame does not work :

    *   for a "perl" older than the 5.8 series ;

    *   for a "DEBUGGING" "perl" run with debugging flags set (as in "perl
        -D ...") ;

    *   when the runloop callback is replaced by another module.

    In those three cases, "uplevel" will look for a "goto &sub" statement in
    its callback and, if there is one, throw an exception before executing
    the code.

    Moreover, in order to handle "goto" statements properly, "uplevel"
    currently has to suffer a run-time overhead proportional to the size of
    the callback in every case (with a small ratio), and proportional to the
    size of all the code executed as the result of the "uplevel" call
    (including subroutine calls inside the callback) when a "goto" statement
    is found in the "uplevel" callback. Despite this shortcoming, this XS
    version of "uplevel" should still run way faster than the pure-Perl
    version from Sub::Uplevel.

    Starting from "perl" 5.19.4, it is unfortunately no longer possible to
    reliably throw exceptions from "uplevel"'d code while the debugger is in
    use. This may be solved in a future version depending on how the core
    evolves.

DEPENDENCIES
    perl 5.6.1.

    A C compiler. This module may happen to build with a C++ compiler as
    well, but don't rely on it, as no guarantee is made in this regard.

    XSLoader (core since perl 5.6.0).

SEE ALSO
    "local" in perlfunc, "Temporary Values via local()" in perlsub.

    Alias, Hook::Scope, Scope::Guard, Guard.

    Sub::Uplevel.

    Continuation::Escape is a thin wrapper around Scope::Upper that gives
    you a continuation passing style interface to "unwind". It's easier to
    use, but it requires you to have control over the scope where you want
    to return.

    Scope::Escape.

AUTHOR
    Vincent Pit, "<perl at profvince.com>", <http://www.profvince.com>.

    You can contact me by mail or on "irc.perl.org" (vincent).

BUGS
    Please report any bugs or feature requests to "bug-scope-upper at
    rt.cpan.org", or through the web interface at

Changes
Revision history for Scope-Upper

0.28    2015-08-18 15:00 UTC
        + Chg : SUB() and EVAL() will now warn if they cannot find an
                appropriate context in the current stack. They will still
                return undef in this case, which is interpreted as the current
                context when combined with other words.
        + Fix : [RT #104751] : Scope::Upper does not handle exotic stack types
                Trying to target a scope above the current perl scope will now
                result in a warning. In that case, the topmost context in the
                current stack will still be returned.
                Thanks Rafaël Garcia-Suarez for the report.
        + Fix : Test failures of threads tests on systems with harsh resource
                constraints causing the threads to exit() during run.
        + Opt : Some internal structures were shrunk, resulting in memory
                savings and small speedups.

0.27    2015-03-27 22:10 UTC
        + Chg : The new environment variable to enable thread tests on older
                perls is PERL_FORCE_TEST_THREADS. Note that this variable
                should only be turned on by authors.
        + Fix : Segfaults when the module is loaded by several threads (or
                Windows emulated processes) ran in parallel.
        + Fix : Memory leak with the uid() feature.
        + Fix : Update the Windows ActivePerl + gcc 3.4 workaround for
                ExtUtils::MakeMaker 7.04. Thanks Christian Walde for reporting
                and feedback on this issue.
        + Fix : reap(), localize(), localize_elem() and localize_delete()
                will again work correctly on perl 5.19.4+ when the debugger
                is enabled.
        + Fix : Silence some compiler warnings.

0.26    2015-03-12 23:30 UTC
        + Fix : [RT #100264] : Don't use CvPADLIST on XSUBs
                Thanks Father Chrysostomos for reporting and contributing a
                patch.
        + Fix : Be really compatible with the optional OP_PARENT feature.

0.25    2014-09-21 17:10 UTC
        + Add : Support for the PERL_OP_PARENT optional feature introduced in
                perl 5.21.2.
        + Fix : Work around an assertion failure in perl 5.21.4.

0.24    2013-09-10 11:10 UTC
        + Fix : Lexicals returned with unwind(), yield() and leave() will no
                longer be lost on perl 5.19.4 and above.

0.23    2013-09-02 11:30 UTC
        This is a maintenance release. The code contains no functional change.
        Satisfied users of version 0.22 can skip this update.
        + Fix : [RT #87178] : typo fixes.
                Thanks dsteinbrunner@pobox.com for the patch.
        + Fix : [RT #88177] : 5.19.3 block hint test breakage
                t/07-context_info.t has been taught about perl 5.19.3.
                Thanks Andrew Main for the patch.
        + Tst : Author tests are no longer bundled with this distribution.
                They are only made available to authors in the git repository.
        + Tst : "given is experimental" warnings when running tests on perl
                5.18 and above have been silenced.

0.22    2013-01-30 23:35 UTC
        + Chg : perl 5.6.1 is required.
        + Fix : context_info() now ignores %warnings::Bits for perl 5.8.6 and
                below, just like caller().
        + Fix : [RT #82220] : Scope-Upper-0.21 test stuck on Windows perl 5.12.3
                t/93-pod-spelling.t will not crash anymore on perl 5.12 and
                below.
                Thanks Gabor Szabo for reporting and Slaven Rezic for debugging
                this issue.
        + Tst : Rare test failures of t/07-context_info.t on perl 5.6,
                t/55-yield-target.t on perl 5.[68], and t/63-uplevel-ctl.t on
                perl 5.8.[345] have been addressed.

0.21    2012-09-21 15:10 UTC
        + Fix : context_info() has been taught about perl 5.17.4.
        + Fix : t/07-context_info.t will no longer fail with perls that have
                sitecustomize enabled (like many perl builds for Windows).
                Thanks Steve Hay and Nicholas Clark for debugging help.
        + Fix : Broken linkage on Windows with gcc 3.4, which appears in
                particular when using ActivePerl's default compiler suite.
                For those setups, the Scope::Upper shared library will now be
                linked against the perl dll directly (instead of the import
                library).

0.20    2012-09-17 11:00 UTC
        + Add : The new yield(@values, $context) function can be used to return
                values to any upper scope, including do or map blocks.
                The new leave(@values) function is an alias for
                yield(@values, HERE).
        + Add : The new context_info($context) function return information
                about context $context, similarly to what caller() provides
                but for any upper scope.
        + Chg : Contexts are now normalized. In previous versions, it was
                possible for different contexts to refer to the same scope :
                for example, "for (my $i = 0; $i < 10; ++$i) { ... }" was
                reachable through two contexts, while "for (@array) { ... }"
                only by one. Starting from this version, contexts are
                normalized so that they always represent an actual scope.
        + Doc : C++ compilers are officially NOT supported.
        + Fix : Building with a more recent version of perl 5.17.4.
        + Fix : Debugger compatibility with perl 5.17.1 and above.

0.19    2012-09-01 13:25 UTC
        + Doc : POD headings have been made linkable.
        + Fix : Building with perl 5.17.4.
        + Fix : BUILD_PREREQS are now set for ExtUtils::MakeMaker.
        + Opt : uplevel() private data structure is 7% smaller on 64 bits
                architectures.
        + Tst : Author tests overhaul.

0.18    2011-10-10 20:50 UTC

Makefile.PL
use 5.006_001;

use strict;
use warnings;
use ExtUtils::MakeMaker;

use Config;

my $pl   = $Config{perl_patchlevel};
my $desc = $Config{git_describe};
for ($pl, $desc) {
 $_ = undef unless defined and length;
}

my @DEFINES;
my %macro;

print "Checking if this is an official release of perl... ";
my $is_release = ("$]" < 5.011) ? (defined($pl) || defined($desc) ? 0 : 1)
                                : (defined($desc)                 ? 0 : 1);
push @DEFINES, "-DSU_RELEASE=$is_release";
print $is_release  ? "yes\n" : "no\n";

my $is_gcc_34 = 0;
print "Checking if this is gcc 3.4 on Windows trying to link against an import library... ";
if ($^O eq 'MSWin32' and not grep /^LD[A-Z]*=/, @ARGV) {
 my ($libperl, $gccversion) = map $_ || '', @Config{qw<libperl gccversion>};
 if ($gccversion =~ /^3\.4\.[0-9]+/ and $libperl =~ s/\.lib$//) {
  $is_gcc_34 = 1;
  my ($lddlflags, $ldflags) = @Config{qw<lddlflags ldflags>};
  $_ ||= '', s/-L(?:".*?"|\S+)//g for $lddlflags, $ldflags;
  $libperl = "-l$libperl";
  my $libdirs = join ' ',
                 map { s/(?<!\\)((?:\\\\)*")/\\$1/g; qq[-L"$_"] }
                  @Config{qw<bin sitebin>};
  $macro{LDDLFLAGS}    = "$lddlflags $libdirs $libperl";
  $macro{LDFLAGS}      = "$ldflags $libdirs $libperl";
  eval <<'  MY_SECTION';
   package MY;
   sub dynamic_lib {
    my $self = shift;
    my $inherited = $self->SUPER::dynamic_lib(@_);
    $inherited =~ s/"?\$\(PERL_ARCHIVE\)"?//g;
    return $inherited;
   }
  MY_SECTION
  die $@ if $@;
 }
}
print $is_gcc_34 ? "yes\n" : "no\n";

# Threads, Windows and 5.8.x don't seem to be best friends
if ($^O eq 'MSWin32' && "$]" < 5.009) {
 push @DEFINES, '-DSU_MULTIPLICITY=0';
}

@DEFINES = (DEFINE => join ' ', @DEFINES) if @DEFINES;
%macro   = (macro  => { %macro })         if %macro; # Beware of the cycle

my $dist = 'Scope-Upper';

(my $name = $dist) =~ s{-}{::}g;

(my $file = $dist) =~ s{-}{/}g;
$file = "lib/$file.pm";

my %PREREQ_PM = (
 'Exporter' => 0,
 'XSLoader' => 0,
 'base'     => 0,
);

my %BUILD_REQUIRES = (
 'ExtUtils::MakeMaker' => 0,
 'Config'              => 0,
 'POSIX'               => 0,
 'Test::More'          => 0,
 %PREREQ_PM,
);

my %META = (
 configure_requires => {
  'Config'              => 0,
  'ExtUtils::MakeMaker' => 0,
 },
 build_requires => {
  %BUILD_REQUIRES,
 },
 dynamic_config => 1,
 resources => {
  bugtracker => "http://rt.cpan.org/Dist/Display.html?Name=$dist",
  homepage   => "http://search.cpan.org/dist/$dist/",
  license    => 'http://dev.perl.org/licenses/',
  repository => "http://git.profvince.com/?p=perl%2Fmodules%2F$dist.git",
 },
);

WriteMakefile(
 NAME             => $name,
 AUTHOR           => 'Vincent Pit <perl@profvince.com>',
 LICENSE          => 'perl',
 VERSION_FROM     => $file,
 ABSTRACT_FROM    => $file,
 PL_FILES         => {},
 @DEFINES,
 BUILD_REQUIRES   => \%BUILD_REQUIRES,
 PREREQ_PM        => \%PREREQ_PM,
 MIN_PERL_VERSION => '5.006001',
 META_MERGE       => \%META,
 dist             => {
  PREOP    => "pod2text -u $file > \$(DISTVNAME)/README",
  COMPRESS => 'gzip -9f', SUFFIX => 'gz'
 },
 clean            => {
  FILES => "$dist-* *.gcov *.gcda *.gcno cover_db Debian_CPANTS.txt*"
 },
 %macro,
);

package MY;

sub postamble {
 return <<'POSTAMBLE';
testdeb: all
	PERL_DL_NONLAZY=1 PERLDB_OPTS="NonStop=1" $(FULLPERLRUN) -MTAP::Harness -e 'TAP::Harness->new({verbosity => q{$(TEST_VERBOSE)}, lib => [ q{$(INST_LIB)}, q{$(INST_ARCHLIB)} ], switches => [ q{-d} ]})->runtests(@ARGV)' $(TEST_FILES)
POSTAMBLE
}
